src/osfiber_asm_ppc64.S

```diff
@@ -0,0 +1,195 @@
+// Copyright 2019 Google LLC
+// Copyright 2019 Shawn Anastasio
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if defined(__powerpc64__)
+
+#define MARL_BUILD_ASM 1
+#include "osfiber_asm_ppc64.h"
+
+// void marl_fiber_swap(marl_fiber_context* from, const marl_fiber_context* to)
+// r3: from
+// r4: to
+.text
+.global marl_fiber_swap
+.align 4
+.type marl_fiber_swap @function
+marl_fiber_swap:
+
+    // Store non-volatile registers
+    std 1, MARL_REG_R1(3)
+    std 2, MARL_REG_R2(3)
+    std 13, MARL_REG_R13(3)
+    std 14, MARL_REG_R14(3)
+    std 15, MARL_REG_R15(3)
+    std 16, MARL_REG_R16(3)
+    std 17, MARL_REG_R17(3)
+    std 18, MARL_REG_R18(3)
+    std 19, MARL_REG_R19(3)
+    std 20, MARL_REG_R20(3)
+    std 21, MARL_REG_R21(3)
+    std 22, MARL_REG_R22(3)
+    std 23, MARL_REG_R23(3)
+    std 24, MARL_REG_R24(3)
+    std 25, MARL_REG_R25(3)
+    std 26, MARL_REG_R26(3)
+    std 27, MARL_REG_R27(3)
+    std 28, MARL_REG_R28(3)
+    std 29, MARL_REG_R29(3)
+    std 30, MARL_REG_R30(3)
+    std 31, MARL_REG_R31(3)
+
+    // Store special registers
+    mflr 5
+    std 5, MARL_REG_LR(3)
+    mfcr 5
+    std 5, MARL_REG_CCR(3)
+
+    // Store non-volatile floating point registers
+    stfd 14, MARL_REG_FPRS+0x10(3)
```

Comment:
Thank you for catching that, you are absolutely correct that this is overflowing the fprs array. I chose to handle fprs and vmx as an array since the non-volatile registers in these classes are contiguous, though I now see how that can lead to issues like this.

For now I'll correct the offsets, but if you would like me to I can unroll the fprs array. The vmx array can also be unrolled, though since it doesn't have displacement-based load/store instructions, I had to stagger load/store and addi instructions which is seems less error prone than manually calculating the offsets like with fprs.

Reply:
I would prefer them unrolled (less cognitive brain power required, and I feel this is more robust for changes on these files), but if you feel strongly I won't block this change just for this.

Reply:
I'll go ahead and unroll the frps array for sure then. If it's okay with you though, I still think that leaving the vmx array as-is is the cleanest way to handle it, especially since vmx registers are 128-bit and representing the registers individually in C would require adding a padding u64 for each register, or creating a separate struct with two u64s to represent a VMX register.

Reply:
>  I still think that leaving the vmx array as-is is the cleanest way to handle it

Yeah, that's fine. I saw they required a bit more special casing.

SGTM, thanks!